File: swap-nodes-in-pairs/swap-nodes-in-pairs.go
// https://leetcode.com/problems/swap-nodes-in-pairs/
// 0 ms 2.1 MB
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func swapPairs(head *ListNode) *ListNode {
	head = &ListNode{
		Next: head,
	}

	for p1 := head; p1 != nil && p1.Next != nil && p1.Next.Next != nil; {
		p1.Next, p1.Next.Next, p1.Next.Next.Next = p1.Next.Next, p1.Next.Next.Next, p1.Next
		p1 = p1.Next.Next
	}

	return head.Next
}

File: swap-nodes-in-pairs/swap-nodes-in-pairs.c
// https://leetcode.com/problems/swap-nodes-in-pairs/
// 0 ms 5.9MB
#include<stdio.h>

struct ListNode {
    int val;
    struct ListNode* next;
};

struct ListNode* swapPairs(struct ListNode* head) {
    struct ListNode pseudo = {.next = head};
    head = &pseudo;
    struct ListNode* p1 = head;
    struct ListNode* p2;
    struct ListNode* p3;

    while (p1 && p1->next && p1->next->next) {
        p2 = p1->next->next;
        p3 = p1->next->next->next;
        p1->next->next->next = p1->next;
        p1->next->next = p3;
        p1->next = p2;
        p1 = p1->next->next;
    }
    return head->next;
}